consumer/consumererror/component.go

```diff
@@ -0,0 +1,38 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import "go.opentelemetry.io/collector/component"
+
+// Component contains information about the component that returned an
+// error. This is helpful when reasoning about the context an error
+// occurred in when handling errors in upstream components within a
+// pipeline.
+type Component struct {
+	err error
+	id  component.ID
```

Comment:
This adds a dependency on `component` in `consumer`, which violates https://github.com/open-telemetry/opentelemetry-collector/issues/9491.

There doesn't appear to be a strong justification for why they are ordered like that, so maybe it's okay. I do think we will need this dependency to effectively communicate this information.

Reply:
We _could_ remove the dependency by making `id` a `string` and then components could set the value using their ID.

Reply:
That breaks features like `ID()`

Reply:
I thought about that, but unless there's some other data we want to put in there, I think all we're doing is making the dependency implicit, which feels worse to me.

Reply:
Agree that using the exactly correct data type is best as long as consumer is allowed to depend on component.

Reply:
> There doesn't appear to be a strong justification for why they are ordered like that, so maybe it's okay. I do think we will need this dependency to effectively communicate this information.

Please don't use "maybe it is okay" when breaking some rules (even if not strongly justified). If it is a rule is a rule, unless we overturn it which is fine but has to be discussed.

Why not have this specific error in the "componenterror" since it is an error about the component? Also how does this interface with `component.Status`?

Reply:
>Please don't use "maybe it is okay" when breaking some rules (even if not strongly justified). If it is a rule is a rule, unless we overturn it which is fine but has to be discussed.

Oh yeah, of course. I just meant that maybe it's okay for us to revise the ordering without violating any design that underpins the current set of rules. I left a comment here as a starting point for discussion.

>Why not have this specific error in the "componenterror" since it is an error about the component? Also how does this interface with component.Status?

I think we will ultimately need this here since it will likely be necessary for retries. It's different from component status since the concepts of errors in pipelines and errors in components are distinct, even if they will occasionally overlap.

I've removed usages of `component.ID` from the design doc for now. We can delay this conversation until we go to tackle retryable errors.